### PR TITLE
Reservation rate excel report

### PR DESCRIPTION
**Pull request by HotStew**
Created capability of downloading the new reservations rate excel report.
Backend PR: https://github.com/andersinno/respa/pull/52

### To be done
- [x]  Tests
- [x]  Making sure project runs on newer node versions
- [x]  Reset state in modal after closing (Elaborated below)

My components somewhat differ from other components in that my have their own state defined within the components. I don't have too much experience with how things are done in this project with the reducers and the selectors so maybe having their own state is wrong?
______________
### Modal location
image
______________
### Unhidden modal with some initial data.
image
______________
### Input some data
image
______________
Clicking download downloads report
image
______________
### Correct filters in excel file
image
image

### CODE REVIEW

File: app/state/reducers/reservationsRateReportModalReducer.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+});
+
+export default handleActions({
+  [uiActions.showReservationsRateReportModal]: state => ({
+    ...state,
+    show: true,
+  }),
+  [uiActions.hideReservationsRateReportModal]: hide,
+  [actionTypes.RESERVATIONS_RATE_REPORT]: () => {},
+}, initialState);
```
Comment (frwickst) [inline app/state/reducers/reservationsRateReportModalReducer.js]: Here you are mixing the use of functions and the settings state in the action handler. I would recommend to pick just one. So either create a `show` function and use that or then refactor the `hide` function here.
The `hide` function could actually be simplified here as well to just `state => initialState`.

File: app/shared/modals/reservation-rate-report/ReservationsRateReport.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    this.props.fetchReservationsRateReport({
+      start_date: this.state.dateStart,
+      end_date: this.state.dateEnd,
+      start_time: this.state.times.begin.time,
+      end_time: this.state.times.end.time,
+      units: this.state.unitSelections.map(selection => selection.value),
+    }).then(() => {
+      this.setState({ loading: false });
+    }).catch(() => {
+      this.setState({ loading: false });
```
Comment (frwickst) [inline app/shared/modals/reservation-rate-report/ReservationsRateReport.js]: There is no error handling here, should there be? Should the user get some kind of message/notification if things did not go as planned?

File: app/shared/modals/reservation-rate-report/ReservationsRateReportContainer.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import Selector from './ReservationsRateReportSelector';
+
+UnconnectedReservationsRateReportModalContainer.propTypes = {
+  onHide: PropTypes.func.isRequired,
+  show: PropTypes.bool.isRequired,
+  units: PropTypes.object.isRequired,
+  fetchReservationsRateReport: PropTypes.func.isRequired,
+};
+
+export function UnconnectedReservationsRateReportModalContainer(props) {
```
Comment (frwickst) [inline app/shared/modals/reservation-rate-report/ReservationsRateReportContainer.js]: Why is this needed? Can't the component be used directly if this is what is needed? As in just using `ReservationsRateReportModal` ?

**Review by frwickst — APPROVED** [review]
Overall this is looking good. I would have prefered having the DateTime component be split up into two different components, since having something called DateTime and then you might only render a Date component can be strange to a developer.
But since I don't know how much refactoring work that would require in this case, I will not require that now.

Just a few things that would be good to check before merging this :)

Good work!

**Review by frwickst — CHANGES_REQUESTED** [review]
wups, wrong selected state by me :D

File: app/shared/modals/reservation-rate-report/ReservationsRateReport.js (hunk 2 above)
Comment (HotStew) [reply app/shared/modals/reservation-rate-report/ReservationsRateReport.js]: Most errors are related to user not selecting correct data in the modal, such as not selecting any units or setting a start date that is after the end date. These are handled on the back end right now though. So should I just implement some local form errors, or just somehow display the error message from the back end?

File: app/shared/modals/reservation-rate-report/ReservationsRateReportContainer.js (hunk 3 above)
Comment (HotStew) [reply app/shared/modals/reservation-rate-report/ReservationsRateReportContainer.js]: Good point, this wrapper is not really needed.

File: app/state/reducers/reservationsRateReportModalReducer.js (hunk 1 above)
Comment (HotStew) [reply app/state/reducers/reservationsRateReportModalReducer.js]: :+1:

File: app/shared/modals/reservation-rate-report/ReservationsRateReport.js (hunk 2 above)
Comment (HotStew) [reply app/shared/modals/reservation-rate-report/ReservationsRateReport.js]: I have decided to go with displaying error messages that we get from the endpoint. I have translated them in the backend. The error message will display on the right side of the modal footer.
Is this solution fine, or should I go with the form route?

**Comment by HotStew** [comment]
I'd also like an opinion/input on the resetting modal state I mentioned in the PR description.

**Comment by HotStew** [comment]
The warning messages below are due to package called `shelljs` which only eslint depends on. This error was introduced when I ran the project on node 14. Apparently it is no big deal. Though, I did try to upgrade eslint 4.x to get rid of the warning messages, but it would have been to much of a hassle because it created a chain effect of updating all the peer dependencies and re-configuring some things which I have no experience of really.

See:
https://github.com/nodejs/node/issues/32987
https://github.com/shelljs/shelljs/pull/973

```
(node:1328200) Warning: Accessing non-existent property 'cat' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
(node:1328200) Warning: Accessing non-existent property 'cd' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'chmod' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'cp' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'dirs' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'pushd' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'popd' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'echo' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'tempdir' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'pwd' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'exec' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'ls' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'find' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'grep' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'head' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'ln' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'mkdir' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'rm' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'mv' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'sed' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'set' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'sort' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'tail' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'test' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'to' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'toEnd' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'touch' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'uniq' of module exports inside circular dependency
(node:1328200) Warning: Accessing non-existent property 'which' of module exports inside circular dependency
```

**Review by frwickst — APPROVED** [review]
LGTM

**Comment by joonvena** [comment]
> I'd also like an opinion/input on the resetting modal state I mentioned in the PR description.

So the problem is that when the modal is closed the previous state inputted there persist instead of resetting the values?

**Review by frwickst — APPROVED** [review]

**Review by frwickst — APPROVED** [review]

**Review by joonvena — APPROVED** [review]

**Review by joonvena — APPROVED** [review]

**Review by frwickst — APPROVED** [review]
👍
